contracts/specs/dai.spec

```diff
@@ -0,0 +1,368 @@
+// dai.spec
+
+// certoraRun contracts/l2/dai.sol:Dai contracts/specs/HashHelper.sol --verify Dai:contracts/specs/dai.spec --rule_sanity --solc_args "['--optimize','--optimize-runs','200']"
+
+using HashHelper as hashHelper
+
+methods {
+    wards(address) returns (uint256) envfree
+    name() returns (string) envfree
+    symbol() returns (string) envfree
+    version() returns (string) envfree
+    decimals() returns (uint8) envfree
+    totalSupply() returns (uint256) envfree
+    balanceOf(address) returns (uint256) envfree
+    allowance(address, address) returns (uint256) envfree
+    nonces(address) returns (uint256) envfree
+    deploymentChainId() returns (uint256) envfree
+    PERMIT_TYPEHASH() returns (bytes32) envfree
+    DOMAIN_SEPARATOR() returns (bytes32) envfree
+    hashHelper.call_ecrecover(bytes32, uint8, bytes32, bytes32) returns (address) envfree
+    hashHelper.computeDigestForDai(bytes32, bytes32, address, address, uint256, uint256, uint256) returns (bytes32) envfree
+}
+
+ghost balanceSum() returns mathint {
+    init_state axiom balanceSum() == 0;
+}
+
+hook Sstore balanceOf[KEY address a] uint256 balance (uint256 old_balance) STORAGE {
+    havoc balanceSum assuming balanceSum@new() == balanceSum@old() + balance - old_balance && balanceSum@new() >= 0;
+}
+
+invariant balanceSum_equals_totalSupply() balanceSum() == totalSupply()
+
```

Comment:
Maybe we could add some more invariants such as address(0) and the contract address never have a balance?